### PR TITLE
[flink] Support dateFormat compute functions

### LINKED ISSUE #1805: [Feature] CDC synchronization supports date Fromat
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

I want to use CDC to write mysql data to paimon and partition it in yyyyMMdd date format

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by fuyun2024**
Support dateFormat compute functions

### Purpose

Linked issue: close #1805

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/Expression.java [hunk 1]
```diff
@@ -37,7 +38,7 @@
 public interface Expression extends Serializable {
 
     List<String> SUPPORTED_EXPRESSION =
-            Arrays.asList("year", "month", "day", "hour", "substring", "truncate");
+            Arrays.asList("year", "month", "day", "hour", "dateFormat", "substring", "truncate");
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/Expression.java:41]: maybe `date_format`?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/Expression.java [hunk 2]
```diff
@@ -239,6 +251,36 @@ public String eval(String input) {
...
+
+        @Override
+        public DataType outputType() {
+            return DataTypes.STRING();
+        }
+
+        @Override
+        public String eval(String input) {
+            LocalDateTime localDateTime = DateTimeUtils.toLocalDateTime(input, 0);
+            return localDateTime.format(DateTimeFormatter.ofPattern(pattern));
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/Expression.java:280]: We should reuse pattern, `DateTimeFormatter.ofPattern(pattern)` could be a class field.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/Expression.java (hunk 1 above)
Comment (fuyun2024) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/Expression.java]: @JingsongLi  I resubmitted the code ， could you help to review this pull request?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/Expression.java [hunk 3]
```diff
@@ -239,6 +251,36 @@ public String eval(String input) {
         }
     }
 
+    /** date format from a time input. */
+    final class DateFormat implements Expression {
+
+        private static final long serialVersionUID = 1L;
+
+        private final String fieldReference;
+        private final DateTimeFormatter formatter;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/Expression.java:260]: Is this Serializable? Maybe you just need a lazy object here.

Comment (fuyun2024) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/Expression.java]: Thank you for your suggestion, i resubmitted the code .

**Pull request merged**